### PR TITLE
feat: Introduce protocol service implementation

### PR DESCRIPTION
**Pull request by soluchok**
* `HandleOutbound` method implementation
* `handle` internal implementation

Issue: https://github.com/hyperledger/aries-framework-go/issues/464

Signed-off-by: Andrii Soluk <isoluchok@gmail.com>

### CODE REVIEW

File: pkg/didcomm/protocol/introduce/service.go [hunk 1]
```diff
@@ -21,6 +23,9 @@ const (
 	Introduce = "introduce"
 	// IntroduceSpec defines the introduce spec
 	IntroduceSpec = metadata.AriesCommunityDID + ";spec/introduce/1.0/"
+	// SkipProposalMsgType defines the skip proposal (the introducer has a public invitation)
+	// skip proposal is not part of protocol specification (internal usage)
+	SkipProposalMsgType = "skip/proposal"
```
Comment (llorllale) [inline pkg/didcomm/protocol/introduce/service.go]: when we spoke about this you mentioned it was going to be some sort of internal implementation detail. Therefore this should definitely not be exported.

Comment (soluchok) [reply pkg/didcomm/protocol/introduce/service.go]: the problem is that there is no way to know that it is skip proposal

Comment (soluchok) [reply pkg/didcomm/protocol/introduce/service.go]: I am replacing it to real one https://github.com/hyperledger/aries-framework-go/pull/516/files#diff-43cf3af5d3c6b9acebc5c635ba4d0f7dR174

Comment (soluchok) [reply pkg/didcomm/protocol/introduce/service.go]: when the client perform this action we need to know somehow that is Skip proposal (and I found the way to create this fake type and then replace it to real one)

Comment (soluchok) [reply pkg/didcomm/protocol/introduce/service.go]: any suggestion welcome :) but I don't think it is a good idea to extend DIDComm message or proposal model just bc we need to know that this is skip proposal. Or we can do some injection to the DB which is the worst solution ever

File: pkg/didcomm/protocol/introduce/service.go [hunk 2]
```diff
@@ -33,10 +40,16 @@ const (
...
+// metaData type to store data for internal usage
+type metaData struct {
+	record
+	Msg      *service.DIDCommMsg
+	ThreadID string
+}
+
+type record struct {
+	StateName string
+	WaitCount int
```
Comment (rolsonquadras) [inline pkg/didcomm/protocol/introduce/service.go]: description about this field would be helpful.

File: pkg/didcomm/protocol/introduce/service.go [hunk 3]
```diff
@@ -127,31 +179,90 @@ func (s *Service) sendMsgEvents(msg service.StateMsg) {
...
 }
 
-func (s *Service) currentState(thID string) (state, error) {
-	name, err := s.store.Get(thID)
+func (s *Service) currentStateRecord(thID string) (*record, error) {
+	src, err := s.store.Get(thID)
+	if errors.Is(err, storage.ErrDataNotFound) {
+		return &record{
+			StateName: stateNameStart,
+			WaitCount: 2,
```
Comment (rolsonquadras) [inline pkg/didcomm/protocol/introduce/service.go]: is the value configurable ?

File: pkg/didcomm/protocol/introduce/service.go [hunk 4]
```diff
@@ -127,31 +179,90 @@ func (s *Service) sendMsgEvents(msg service.StateMsg) {
...
+	if err := json.Unmarshal(src, &r); err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+func (s *Service) save(id string, data interface{}) error {
+	src, err := json.Marshal(data)
+	if err != nil {
+		return fmt.Errorf("failed to mashal: %w", err)
```
Comment (rolsonquadras) [inline pkg/didcomm/protocol/introduce/service.go]: https://github.com/hyperledger/aries-framework-go/pull/525#discussion_r337080113 applicable here.

File: pkg/didcomm/protocol/introduce/service.go [hunk 5]
```diff
@@ -187,8 +298,62 @@ func canTriggerActionEvents(msg *service.DIDCommMsg) bool {
...
+			WaitCount: msg.WaitCount,
+		}); err != nil {
+			return fmt.Errorf("failed to persist state %s %w", next.Name(), err)
+		}
+
+		logger.Infof("persisted the connection using %s and updated the state to %s", msg.ThreadID, next.Name())
+
+		s.sendMsgEvents(service.StateMsg{
+			Type: service.PostState, Msg: msg.Msg, StateID: next.Name(),
+			//Properties: createEventProperties(msg.ThreadID, ""),
```
Comment (rolsonquadras) [inline pkg/didcomm/protocol/introduce/service.go]: Can we remove the commented code ? Better to have a clean code in master.

File: pkg/didcomm/protocol/introduce/service.go (hunk 3 above)
Comment (soluchok) [reply pkg/didcomm/protocol/introduce/service.go]: probably in the future :) I found TODO in spec
`TODO:do we want to support introducing more than 2 at a time?`

**Pull request merged**